Adds support for listing tags, filtering machines by tag, and getting machine hardware info

markylaing opened the pull request:
* Adds a new interface `Tag` (implemented by `tag` in `tags.go`) for interacting with MAAS tags.
* Adds a method `Tags() ([]Tag, error)` to the `Controller` for retrieving a slice of MAAS tags.
* Adds `Tags` to `MachineArgs` to allow filtering machines by tag.
* Adds a method `HardwareInfo() map[string]string` to the `Machine` interface and implements it. Hardware info is added to `machine` when deserializing the API response.

I have tried to follow conventions that I have found elsewhere in the codebase but let me know if I have missed something, thanks :smile:

markylaing committed:
tags: Adds and implements Tag interface.

Adds required methods for reading a list of tags.

Signed-off-by: Mark Laing <mark.laing@canonical.com>

markylaing committed:
controller: Adds and implements Tags method on controller.

Adds method to controller interface for listing all MAAS tags.

Signed-off-by: Mark Laing <mark.laing@canonical.com>

markylaing committed:
controller: Filter machines by tag.

Adds Tags field to MachinesArgs and adds these values to query parameters
to filter results.

Signed-off-by: Mark Laing <mark.laing@canonical.com>

markylaing committed:
machine: Adds and implements HardwareInfo on Machine interface.

Adds methods for retrieval of hardware info from the MAAS API response
for a machine.

Signed-off-by: Mark Laing <mark.laing@canonical.com>

stgraber commented on tags.go:
```diff
@@ -0,0 +1,124 @@
+// Copyright 2016 Canonical Ltd.
```
Probably ought to say 2022 :)

markylaing replied:
Fixed.

jameinel commented:
Tags / MachineArgs and HardwareInfo seem relatively straightforward, but it is a bit of a surprise that you're also changing how dispatchSingleRequest (and adding an HTTPClient override), without document the what and why you're changing it in the PR overview.
(It also seems weird to me that dispatchSingleRequest is not a pointer receiver and spawns an http.Client for every request, rather than allowing connection sharing.)
can you elaborate a bit on those changes?

stgraber commented:
@markylaing is out for a few days but I agree that it'd make sense to clarify the need for the http.Client in the PR or maybe even split this into two PRs.

The reason why we want to have a custom http.Client is so we can do proper TLS validation of otherwise self-signed servers as well as more tightly control timeouts.

jameinel commented on machine.go:
```diff
@@ -574,6 +587,13 @@ func machine_2_0(source map[string]interface{}) (*machine, error) {
 		return nil, errors.Trace(err)
 	}
+
+	validHardwareInfo := valid["hardware_info"].(map[string]interface{})
```
Is this safe to do? If it doesn't have the right type this will be a runtime panic. I'd be more comfortable with the:

validHardwareInfo, ok := valid["hardware_info"].(map[string]interface{})
if !ok {
  // not actually a map[string]interface{}
}

markylaing replied:
Happy to add these checks. My understanding here was that coercion fails if `hardware_info` is not a `map[string]string` as defined by the schema above. Is it possible for MAAS to omit the `hardware_info` field or return a different type?

markylaing replied:
I've added these checks now.

jameinel commented on machine.go:
```diff
@@ -574,6 +587,13 @@ func machine_2_0(source map[string]interface{}) (*machine, error) {
+	validHardwareInfo := valid["hardware_info"].(map[string]interface{})
+	hardwareInfo := make(map[string]string, len(validHardwareInfo))
+	for k, v := range validHardwareInfo {
+		hardwareInfo[k] = v.(string)
```
Same here. Runtime panics break a lot of things, so probably better to do a
```
val, ok := v.(string)
if ok {
  hardwareInfo[k] = val
} else {
  // Log a warning about data not being in the format you expected
}
```

It is likely to be one of those "should never happen" places, but when it isn't you end up panicing and having a hard time understanding what data is actually invalid, etc.

markylaing replied:
I've added a check for this now, it returns an error instead of panicking.

jameinel commented on tags.go:
```diff
@@ -0,0 +1,124 @@
+		return nil, WrapWithDeserializationError(err, "machine base schema check failed")
+	}
+
+	valid := coerced.([]interface{})
```
This is reasonable because you did just do a schema validation, but it is probably worth adding a comment to this line to indicate that fact. `// ok to do a direct cast because we just coerced the interface`

jameinel commented on tags.go:
```diff
@@ -0,0 +1,124 @@
+
+func getTagDeserializationFunc(controllerVersion version.Number) (tagDeserializationFunc, error) {
+	var deserialisationVersion version.Number
+	for v := range tagDeserializationFuncs {
```
It seems a bit odd to have a map/list here but only 1 entry in the list.

markylaing replied:
For this I followed the pattern found in:
* https://github.com/juju/gomaasapi/blob/v2/link.go#L85
* https://github.com/juju/gomaasapi/blob/v2/interface.go#L372
* https://github.com/juju/gomaasapi/blob/v2/machine.go#L497

I had assumed this was for compatibility with later API versions but happy to remove.

jameinel replied:
It does seem like a standard pattern. Fair enough. (I think at one point we probably were thinking about MAAS 1.X and potential 3.x)

jameinel reviewed (commented):
If you did want to expose httpClient, we would definitely want to add some tests that it is supported/being used when appropriate.

markylaing commented:
> @markylaing is out for a few days but I agree that it'd make sense to clarify the need for the http.Client in the PR or maybe even split this into two PRs.
> 
> The reason why we want to have a custom http.Client is so we can do proper TLS validation of otherwise self-signed servers as well as more tightly control timeouts.

@jameinel @stgraber Ok I'll split this into two PRs.

markylaing committed:
machine: Adds hardware info to test data.

Signed-off-by: Mark Laing <mark.laing@canonical.com>

markylaing committed:
tags: Adds tests for reading tags.

Signed-off-by: Mark Laing <mark.laing@canonical.com>

markylaing commented:
@jameinel @stgraber I think I've addressed all of your comments. Here is a link to the second PR containing the commits which allow a configurable http client: https://github.com/juju/gomaasapi/pull/97

jameinel commented on controller.go:
```diff
@@ -1056,3 +1058,24 @@ func convertConstraintMatchesAny(source interface{}) map[string][]interface{} {
+// Tags implements Controller.
+func (c *controller) Tags() ([]Tag, error) {
+	params := NewURLParams()
+	source, err := c.getQuery("tags", params.Values)
```
It seems a little odd that the way to do the controller tags to pass an empty url.Values, but I trust you to have studied the MAAS api itself for this.

markylaing replied:
You are right - this should just be `nil`. Fixed and thanks!

jameinel commented:
/build

jameinel commented:
Failure on this build was because of a misconfiguration of Jenkins. I'm looking to fix that and will resubmit.

markylaing commented:
@jameinel Did you have any success fixing the Jenkins configuration?

jameinel commented:
Sorry about the delay, I was off on holiday. I will try again today.

John
=:->


On Tue, Jun 21, 2022 at 7:35 AM Mark Laing ***@***.***> wrote:

> @jameinel <https://github.com/jameinel> Did you have any success fixing
> the Jenkins configuration?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/juju/gomaasapi/pull/96#issuecomment-1161626425>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRQ7PFIYHQWOGJI5KWWA3VQGSHVANCNFSM5XBCVI5A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

markylaing commented:
@jameinel no worries :+1:

jameinel commented:
/build

jameinel commented:
/build

jameinel commented:
It has been fixed, and the build check passed. So I'll merge.

jameinel commented:
/merge

jameinel commented:
/merge

markylaing commented:
> It has been fixed, and the build check passed. So I'll merge.

@jameinel thank you!
